jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceListing.java

```diff
@@ -200,57 +210,125 @@ public static String getAsHTML(Resource resource, String base, boolean parent, S
 
         DateFormat dfmt = DateFormat.getDateTimeInstance(DateFormat.MEDIUM, DateFormat.MEDIUM);
 
-        for (Resource item : listing)
+        for (Path item : listing)
         {
-            // TODO this feels fragile, as collections probably should not return a Path here
-            //      and even if they do, it might not be named correctly
-            String name = item.getPath().toFile().getName();
+            Path fileName = item.getFileName();
+            if (fileName == null)
+            {
+                continue; // skip
+            }
+
+            String name = fileName.toString();
             if (StringUtil.isBlank(name))
-                continue;
+            {
+                return null;
+            }
 
-            if (item.isDirectory() && !name.endsWith("/"))
+            if (Files.isDirectory(item))
+            {
                 name += URIUtil.SLASH;
+            }
 
             // Name
             buf.append("<tr><td class=\"name\"><a href=\"");
-            // TODO should this be a relative link?
-            String path = URIUtil.addEncodedPaths(encodedBase, URIUtil.encodePath(name));
-            buf.append(path);
+
+            String href = URIUtil.addEncodedPaths(encodedBase, URIUtil.encodePath(name));
+            buf.append(href);
             buf.append("\">");
             buf.append(deTag(name));
-            buf.append("&nbsp;");
+            buf.append(NBSP);
             buf.append("</a></td>");
 
             // Last Modified
             buf.append("<td class=\"lastmodified\">");
-            long lastModified = item.lastModified();
-            if (lastModified > 0)
-                buf.append(dfmt.format(new Date(item.lastModified())));
-            buf.append("&nbsp;</td>");
+
+            try
+            {
+                FileTime lastModified = Files.getLastModifiedTime(item, LinkOption.NOFOLLOW_LINKS);
+                buf.append(dfmt.format(new Date(lastModified.toMillis())));
+            }
+            catch (IOException ignore)
+            {
+                // do nothing (lastModifiedTime not supported by this file system)
+            }
```

Comment:
This is exactly the kind of boiler plate that can be hidden inside a `Resource` class.  Besides, I assume the lastModifiedTime for a resource collection would be the latest date of all its contained resources, so again this fails because it assumes there is a 1:1 between resource and path.

Reply:
This doesn't assume ResourceCollection vs Resource, this is guaranteed to always be a Path at this point.

Reply:
There can be more than one path at this point. If multiple items in a resource collection contain the same sub directory, then the listed item should be a resource collection and the last modified date should be the most recent of its items.

Reply:
I agree with @gregw on the boilerplate. You don't want to remember, every single time you need a last modified time, to use the right `LinkOption` and lookup the date formatting APIs (why are we not using `java.time` here, but the old `java.text`?).

I also feel that the compositeness of `Resource` is still necessary, and would reduce other boilerplate.
For example, we don't want to rewrite all the times a recursive `collectListing()` (should it not be just shallow for directory listing?) every time we need children resources.

To me, `list()` has the wrong semantic. It should be `resource.getChildren()` and, only if really needed, resource.getDescendants()`, which are easily implemented in both the "single" case and the "composite" case.
This would make crystal clear that `Resource` has a tree structure, and to display a directory listing you obviously want to call `getChildren()` (not `getDescendants()`), and then do the sorting externally (as dictated by the requirements of the directory listing request, e.g. query parameters).

Reply:
Ok, several things being discussed here.

_First, `lastModified` ..._

What would you see the API as?
`Resource.getLastModifiedMillis()` ?

The formatting of millis to date was not updated in this PR, this is the technique used since Jetty 7.0.0

See: https://github.com/eclipse/jetty.project/commit/456a00161b47b8a3a13fc253c3a94ccacd69ea3b#diff-42810570fa3d8b562ad62d94a4ae857667aeb10b17f4ef9e93af43411af3d0aaR471

If we want to use the new `java.time` APIs, this would be a good time to update that.

_Next, the proposed `Resource` listing APIs ..._

`Resource.getChildren()` returns what?
I would hope it's `Stream<Path>` as that's going to produce TONS less garbage than `List<Resource>` (and I can prove it in a jmh test).
Even if the API is `Stream<Resource>` it will produce less garbage overall when it's used.
Even the implementation on `ResourceCollection.getChildren()` could handle that.

If we cared about garbage, then the alias handling would be 100% optional and only trigger in the code paths that require it (currently only things using ResourceService or DefaultServlet).  It should not trigger for Configuration handling, or ResourceListing, or ServletContext.getResource(String).
It would have been a better solution to have `Alias Resource.getAlias()` as the only API for alias on `Resource`, and it only executes when asked for, returning the information needed by AliasChecker to evaluate if the resource is allowed or not.

The `Resource.getDescendants()` is not a good API for the modern era with webapps and/or resourceBases that have 10,000+ entries (super common, especially with spring-boot applications that use javascript libs).  It made sense in the past, but not today.

As a replacement for `Resource.getDescendants()`, how about a `List<Resource> Resource.find(Predicate predicate)` and let the implementations find what is being asked for.

_Next, the concept of a Resource a as a tree structure._

This is the source of many bugs, even in Jetty 7 thru Jetty 11.
We shouldn't have hidden `ResourceCollection` behind the `Resource` interface.
If the API can, possible, even remotely, be a collection, then it should indicate that in it's API, use the ResourceCollection object, not Resource.

Take, for example, the `Resource WebAppContext.getWebInf()`.
This API, at face value, can be a `ResourceCollection`, but in practice, it's assumed to not be one in many places in the Configuration flows.
Take this simple example, `EnvConfiguration` and it's `jetty-env.xml` (there are more if you look, this one is just easy to explain).
If we have a `ResourceCollection` with multiple `WEB-INF` directory entries (as returned by `WebAppContext.getWebInf()`), then it has the potential for returning multiple `jetty-env.xml` entries.    This entry is eventually passed to the `XmlConfiguration(Resource)` to load the XML.  `XmlConfiguration(Resource)` does not support `ResourceCollection` objects.

The current `ResourceCollection.resolve("jetty-env.xml")` punts on this and only returns the first one.   
The user has to be aware of this and ensure their collection order is sane for it to function in a predictable way.
But in reality, the `jetty-env.xml` is now a composite of multiple `jetty-env.xml` entries found potentially in many places. (just like web fragments)

This means `ResourceCollection.resolve(String)` should not return "first hit", but "all hits" in this case.
Or we should have a `ResourceColllection ResourceCollection.resolveAll(String)`, but that means being `ResourceCollection` aware again, as `Resource.resolveAll(String)` doesn't make sense.

When working with `Resource` and the possibility of if being a `ResourceCollection` and having to access the `Resource` it represents (read contents, last modified, size, etc) the API breaks again.   You have to unwrap the `Resource` if it's a `ResourceCollection` to get those APIs to function properly.

Reply:
This is no different to Path itself.  It can represent either a directory or a file (or other things).  If you use API in the wrong way with the wrong Path it "breaks". ie listing entries for a file or asking for the length of a directory.  This is nothing new.  It is a duality with File Descriptors, Files, Paths and Resources.    It is a very common pattern we just have to live  with.